### PR TITLE
Caffe-plugin: 1) add listoutputs 2) fix code style & warning

### PR DESCRIPTION
**Pull request by HrWangChengdu**

### CODE REVIEW

File: plugin/caffe/caffe_loss-inl.h [hunk 1]
```diff
@@ -141,7 +141,7 @@ class CaffeLoss : public Operator {
     top_[0]->set_cpu_diff(&grad_scale_);
 
     // Set BP flag
-    for (index_t i = 0; i < param_.num_data; ++i)
+    for (int i = 0; i < param_.num_data; ++i)
       flags_[i] = req[i] != kNullOp;
 
     caffeOp_->Backward(top_, flags_, bot_);
```
Comment (piiswrong) [inline plugin/caffe/caffe_loss-inl.h]: @HrWangChengdu Please check if you need to call get_cpu/gpu_diff for all the output blobs. Sometimes caffe gpu layers fallback to cpu and the output is only copied to gpu when the next layer calls get_gpu_data/diff.
Same for forward.

Comment (HrWangChengdu) [inline plugin/caffe/caffe_loss-inl.h]: @piiswrong I checked all layers with any cpu operation in gpu code and indeed found one case, i.e. forward_gpu() of ReductionLayer, that needs to call gpu_data/diff to sync.
I'll add cpu->gpu data/diff sync at the end of both forward and backward.

**Comment by HrWangChengdu** [comment]
@piiswrong Please check https://github.com/dmlc/mxnet/pull/3017

**Comment by HrWangChengdu** [comment]
@piiswrong I need to sleep now... will work on caffe's learning rate stuff tomorrow.

**Comment by HrWangChengdu** [comment]
@piiswrong I replaced lenet by mxnet's op in caffe/example, training is fine under cpu and gpu mode in my machine.

**Comment by piiswrong** [comment]
are you sure? I and another guy both met the problem. Here is the symbol we used:
```
def get_mine():
    data = mx.symbol.Variable('data')
    
    conv1 = mx.symbol.Convolution(name='conv1', data=data , num_filter=20, pad=(0,0), kernel=(5,5), stride=(1,1), no_bias=False)
    pool1 = mx.symbol.Pooling(name='pool1', data=conv1 , pad=(0,0), kernel=(2,2), stride=(2,2), pool_type='max')
    conv2 = mx.symbol.Convolution(name='conv2', data=pool1 , num_filter=50, pad=(0,0), kernel=(5,5), stride=(1,1), no_bias=False)
    pool2 = mx.symbol.Pooling(name='pool2', data=conv2 , pad=(0,0), kernel=(2,2), stride=(2,2), pool_type='max')
    flatten_0=mx.symbol.Flatten(name='flatten_0', data=pool2)
    ip1 = mx.symbol.FullyConnected(name='ip1', data=flatten_0 , num_hidden=500, no_bias=False)
    relu1 = mx.symbol.Activation(name='relu1', data=ip1 , act_type='relu')
    ip2 = mx.symbol.FullyConnected(name='ip2', data=relu1 , num_hidden=10, no_bias=False)
    softmax = mx.symbol.SoftmaxOutput(name='softmax', data=ip2 )
    
    return softmax;
```

**Comment by HrWangChengdu** [comment]
Haha. I know why. In get_mine(), activation layers between conv and pool are missed. Then the original learning rate could be too big for the network to train. However, if you add activ layers, then training is fine.

**Comment by piiswrong** [comment]
Ok. then it's just network difference.

**Pull request merged**